src/jaxsim/rbda/contacts/soft.py

```diff
@@ -157,179 +200,161 @@ class SoftContacts(ContactModel):
+        # in a terrain area with a locally constant normal. However, this assumption
+        # is not true in general for highly uneven terrains.
+        m_normal = jnp.dot(m, n̂) * n̂
+        m_tangential = m - jnp.dot(m, n̂) * n̂
+
+        # Compute the tangential force in the sticking case.
+        f_tangential = -((K * δp) * m_tangential + (D * δq) * v_tangential)
+
+        # Detect the contact type (sticking or slipping).
+        # Note that if there is no contact, sticking is set to True, and this detail
+        # is exploited in the computation of the `contact_status` variable.
+        sticking = jnp.logical_or(
+            δ <= 0, f_tangential.dot(f_tangential) <= (μ * force_normal_mag) ** 2
+        )
 
-        # Compute the penetration normal velocity.
-        δ̇ = -jnp.dot(W_ṗ_C, n̂)
+        # Compute the direction of the tangential force.
+        # To prevent dividing by zero, we use a switch statement.
+        # The ε, instead, is needed to make AD happy.
+        f_tangential_direction = jnp.where(
+            f_tangential.dot(f_tangential) != 0,
+            f_tangential / jnp.linalg.norm(f_tangential + ε),
+            jnp.zeros(3),
+        )
 
-        # Non-linear spring-damper model.
-        # This is the force magnitude along the direction normal to the terrain.
-        force_normal_mag = jax.lax.select(
-            pred=δ >= 1e-9,
-            on_true=jnp.sqrt(δ + 1e-12) * (K * δ + D * δ̇),
-            on_false=jnp.array(0.0),
+        # Project the tangential force to the friction cone if slipping.
+        f_tangential = jnp.where(
+            sticking,
+            f_tangential,
+            jnp.minimum(μ * force_normal_mag, jnp.linalg.norm(f_tangential + ε))
+            * f_tangential_direction,
         )
 
-        # Prevent negative normal forces that might occur when δ̇ is largely negative.
-        force_normal_mag = jnp.maximum(0.0, force_normal_mag)
+        # Set the tangential force to zero if there is no contact.
+        f_tangential = jnp.where(δ <= 0, jnp.zeros(3), f_tangential)
+
+        # =====================================
+        # Compute the material deformation rate
+        # =====================================
+
+        # Compute the derivative of the material deformation.
+        ṁ_no_contact = -(K / D) * m
+        ṁ_sticking = v_tangential - (K / D) * m_normal
+        ṁ_slipping = -(f_tangential + (K * δp) * m_tangential) / (D * δq)
+
+        # Compute the contact status:
+        # 0: slipping
+        # 1: sticking
+        # 2: no contact
+        contact_status = sticking.astype(int)
+        contact_status += (δ <= 0).astype(int)
+
+        # Select the right material deformation rate depending on the contact status.
+        ṁ = jax.lax.switch(
+            index=contact_status,
+            branches=(lambda: ṁ_slipping, lambda: ṁ_sticking, lambda: ṁ_no_contact),
+        )
```

Comment:
We could also use `jax.lax.select_n` if it is more sound to you
```suggestion
        ṁ = jax.lax.select_n(
            ṁ_slipping, ṁ_sticking, ṁ_no_contact,
            which=contact_status,
        )
```